File: packages/gbnf-experiment/data/20260909T185402Z_14873f98/ported_implementation/gbnf/grammar_graph/get_parent_stack_id.py
from .colorize import Color


def get_parent_stack_id(pointer, col) -> str:
    stack_ids: list[str] = []
    parent = pointer.parent
    while parent is not None:
        meta = parent.node.meta
        stack_ids.append(f"{meta.stack_id},{meta.path_id},{meta.step_id}")
        parent = parent.parent
    return col("<-", Color.GRAY).join(col(i, Color.RED) for i in stack_ids)

File: proposed-projects/round-trip-experiment/reverse/20260909T190508Z_7c1544c0/20260910T103012Z_7e3d5183/ported_implementation/src/grammar-graph/get-parent-stack-id.ts
import { Color, type Colorize } from './colorize.js';
import type { GraphPointer } from './graph-pointer.js';

export const getParentStackId = (
  pointer: GraphPointer,
  col: Colorize
): string => {
  const stackIds: string[] = [];
  let parent = pointer.parent;
  while (parent) {
    const meta = parent.node.meta;
    stackIds.push(`${meta.stackId},${meta.pathId},${meta.stepId}`);
    parent = parent.parent;
  }
  return stackIds
    .map((id) => col(id, Color.RED))
    .join(col('<-', Color.GRAY));
};
